```
15:17:44 [ERROR] Failures: 
15:17:44 [ERROR]   UserApiTest.testUserManagement:67 expected:<...erSize": 5
15:17:44   },
15:17:44   "a[]pplications": [ ],
15:17:44  ...> but was:<...erSize": 5
15:17:44   },
15:17:44   "a[rchiveAccess": { },
15:17:44   "a]pplications": [ ],
15:17:44  ...>
```

--- a/controller-api/src/main/java/com/yahoo/vespa/hosted/controller/api/role/PathGroup.java
+++ b/controller-api/src/main/java/com/yahoo/vespa/hosted/controller/api/role/PathGroup.java
@@ -62,7 +62,10 @@ enum PathGroup {
                "/application/v4/tenant/{tenant}/key/"),
 
     tenantArchiveAccess(Matcher.tenant,
-                       "/application/v4/tenant/{tenant}/archive-access"),
+                       "/application/v4/tenant/{tenant}/archive-access",
+                       "/application/v4/tenant/{tenant}/archive-access/aws",
+                       "/application/v4/tenant/{tenant}/archive-access/gcp"),
+
 
     billingToken(Matcher.tenant,
                  "/billing/v1/tenant/{tenant}/token"),

--- a/controller-api/src/main/java/com/yahoo/vespa/hosted/controller/tenant/ArchiveAccess.java
+++ b/controller-api/src/main/java/com/yahoo/vespa/hosted/controller/tenant/ArchiveAccess.java
@@ -24,7 +24,7 @@ public class ArchiveAccess {
         this(Optional.empty(), Optional.empty());
     }
 
-    public ArchiveAccess(Optional<String> awsRole, Optional<String> gcpMember) {
+    private ArchiveAccess(Optional<String> awsRole, Optional<String> gcpMember) {
         this.awsRole = awsRole;
         this.gcpMember = gcpMember;
 
@@ -32,12 +32,28 @@ public class ArchiveAccess {
         gcpMember.ifPresent(member -> validateGCPMember(member));
     }
 
-    public static ArchiveAccess fromAWSRole(String role) {
-        return new ArchiveAccess(Optional.of(role), Optional.empty());
+    public ArchiveAccess withAWSRole(String role) {
+        return new ArchiveAccess(Optional.of(role), gcpMember());
     }
 
-    public static ArchiveAccess fromGCPMember(String member) {
-        return new ArchiveAccess(Optional.empty(), Optional.of(member));
+    public ArchiveAccess withGCPMember(String member) {
+        return new ArchiveAccess(awsRole(), Optional.of(member));
+    }
+
+    public ArchiveAccess withAWSRole(Optional<String> role) {
+        return new ArchiveAccess(role, gcpMember());
+    }
+
+    public ArchiveAccess withGCPMember(Optional<String> member) {
+        return new ArchiveAccess(awsRole(), member);
+    }
+
+    public ArchiveAccess removeAWSRole() {
+        return new ArchiveAccess(Optional.empty(), gcpMember());
+    }
+
+    public ArchiveAccess removeGCPMember() {
+        return new ArchiveAccess(awsRole(), Optional.empty());
     }
 
     private void validateAWSIAMRole(String role) {

--- a/controller-server/src/main/java/com/yahoo/vespa/hosted/controller/persistence/TenantSerializer.java
+++ b/controller-server/src/main/java/com/yahoo/vespa/hosted/controller/persistence/TenantSerializer.java
@@ -210,7 +210,7 @@ public class TenantSerializer {
         // TODO(enygaard, 2022-05-24): Remove when all tenants have been rewritten to use ArchiveAccess object
         Optional<String> archiveAccessRole = SlimeUtils.optionalString(tenantObject.field(archiveAccessRoleField));
         if (archiveAccessRole.isPresent()) {
-            return new ArchiveAccess(archiveAccessRole, Optional.empty());
+            return new ArchiveAccess().withAWSRole(archiveAccessRole.get());
         }
         Inspector object = tenantObject.field(archiveAccessField);
         if (!object.valid()) {
@@ -218,7 +218,9 @@ public class TenantSerializer {
         }
         Optional<String> awsArchiveAccessRole = SlimeUtils.optionalString(object.field(awsArchiveAccessRoleField));
         Optional<String> gcpArchiveAccessMember = SlimeUtils.optionalString(object.field(gcpArchiveAccessMemberField));
-        return new ArchiveAccess(awsArchiveAccessRole, gcpArchiveAccessMember);
+        return new ArchiveAccess()
+                .withAWSRole(awsArchiveAccessRole)
+                .withGCPMember(gcpArchiveAccessMember);
     }
     TenantInfo tenantInfoFromSlime(Inspector infoObject) {
         if (!infoObject.valid()) return TenantInfo.empty();

--- a/controller-server/src/main/java/com/yahoo/vespa/hosted/controller/restapi/application/ApplicationApiHandler.java
+++ b/controller-server/src/main/java/com/yahoo/vespa/hosted/controller/restapi/application/ApplicationApiHandler.java
@@ -297,7 +297,9 @@ public class ApplicationApiHandler extends AuditLoggingRequestHandler {
         if (path.matches("/application/v4/tenant/{tenant}/access/approve/operator")) return approveAccessRequest(path.get("tenant"), request);
         if (path.matches("/application/v4/tenant/{tenant}/access/managed/operator")) return addManagedAccess(path.get("tenant"));
         if (path.matches("/application/v4/tenant/{tenant}/info")) return updateTenantInfo(path.get("tenant"), request);
-        if (path.matches("/application/v4/tenant/{tenant}/archive-access")) return allowArchiveAccess(path.get("tenant"), request);
+        if (path.matches("/application/v4/tenant/{tenant}/archive-access")) return allowAwsArchiveAccess(path.get("tenant"), request); // TODO(enygaard, 2022-05-25) Remove when no longer used by console
+        if (path.matches("/application/v4/tenant/{tenant}/archive-access/aws")) return allowAwsArchiveAccess(path.get("tenant"), request);
+        if (path.matches("/application/v4/tenant/{tenant}/archive-access/gcp")) return allowGcpArchiveAccess(path.get("tenant"), request);
         if (path.matches("/application/v4/tenant/{tenant}/secret-store/{name}")) return addSecretStore(path.get("tenant"), path.get("name"), request);
         if (path.matches("/application/v4/tenant/{tenant}/application/{application}/instance/{instance}/environment/{environment}/region/{region}/global-rotation/override")) return setGlobalRotationOverride(path.get("tenant"), path.get("application"), path.get("instance"), path.get("environment"), path.get("region"), false, request);
         if (path.matches("/application/v4/tenant/{tenant}/application/{application}/environment/{environment}/region/{region}/instance/{instance}/global-rotation/override")) return setGlobalRotationOverride(path.get("tenant"), path.get("application"), path.get("instance"), path.get("environment"), path.get("region"), false, request);
@@ -345,7 +347,9 @@ public class ApplicationApiHandler extends AuditLoggingRequestHandler {
         if (path.matches("/application/v4/tenant/{tenant}")) return deleteTenant(path.get("tenant"), request);
         if (path.matches("/application/v4/tenant/{tenant}/access/managed/operator")) return removeManagedAccess(path.get("tenant"));
         if (path.matches("/application/v4/tenant/{tenant}/key")) return removeDeveloperKey(path.get("tenant"), request);
-        if (path.matches("/application/v4/tenant/{tenant}/archive-access")) return removeArchiveAccess(path.get("tenant"));
+        if (path.matches("/application/v4/tenant/{tenant}/archive-access")) return removeAwsArchiveAccess(path.get("tenant"));  // TODO(enygaard, 2022-05-25) Remove when no longer used by console
+        if (path.matches("/application/v4/tenant/{tenant}/archive-access/aws")) return removeAwsArchiveAccess(path.get("tenant"));
+        if (path.matches("/application/v4/tenant/{tenant}/archive-access/gcp")) return removeGcpArchiveAccess(path.get("tenant"));
         if (path.matches("/application/v4/tenant/{tenant}/secret-store/{name}")) return deleteSecretStore(path.get("tenant"), path.get("name"), request);
         if (path.matches("/application/v4/tenant/{tenant}/application/{application}")) return deleteApplication(path.get("tenant"), path.get("application"), request);
         if (path.matches("/application/v4/tenant/{tenant}/application/{application}/deployment")) return removeAllProdDeployments(path.get("tenant"), path.get("application"));
@@ -1028,7 +1032,7 @@ public class ApplicationApiHandler extends AuditLoggingRequestHandler {
         return new SlimeJsonResponse(slime);
     }
 
-    private HttpResponse allowArchiveAccess(String tenantName, HttpRequest request) {
+    private HttpResponse allowAwsArchiveAccess(String tenantName, HttpRequest request) {
         if (controller.tenants().require(TenantName.from(tenantName)).type() != Tenant.Type.cloud)
             throw new IllegalArgumentException("Tenant '" + tenantName + "' is not a cloud tenant");
 
@@ -1036,27 +1040,62 @@ public class ApplicationApiHandler extends AuditLoggingRequestHandler {
         var role = mandatory("role", data).asString();
 
         if (role.isBlank()) {
-            return ErrorResponse.badRequest("Archive access role can't be whitespace only");
+            return ErrorResponse.badRequest("AWS archive access role can't be whitespace only");
         }
 
         controller.tenants().lockOrThrow(TenantName.from(tenantName), LockedTenant.Cloud.class, lockedTenant -> {
-            lockedTenant = lockedTenant.withArchiveAccess(new ArchiveAccess(Optional.of(role), Optional.empty()));
+            var access = lockedTenant.get().archiveAccess();
+            lockedTenant = lockedTenant.withArchiveAccess(access.withAWSRole(role));
             controller.tenants().store(lockedTenant);
         });
 
-        return new MessageResponse("Archive access role set to '" + role + "' for tenant " + tenantName +  ".");
+        return new MessageResponse("AWS archive access role set to '" + role + "' for tenant " + tenantName +  ".");
     }
 
-    private HttpResponse removeArchiveAccess(String tenantName) {
+    private HttpResponse removeAwsArchiveAccess(String tenantName) {
         if (controller.tenants().require(TenantName.from(tenantName)).type() != Tenant.Type.cloud)
             throw new IllegalArgumentException("Tenant '" + tenantName + "' is not a cloud tenant");
 
         controller.tenants().lockOrThrow(TenantName.from(tenantName), LockedTenant.Cloud.class, lockedTenant -> {
-            lockedTenant = lockedTenant.withArchiveAccess(new ArchiveAccess());
+            var access = lockedTenant.get().archiveAccess();
+            lockedTenant = lockedTenant.withArchiveAccess(access.removeAWSRole());
             controller.tenants().store(lockedTenant);
         });
 
-        return new MessageResponse("Archive access role removed for tenant " + tenantName + ".");
+        return new MessageResponse("AWS archive access role removed for tenant " + tenantName + ".");
+    }
+
+    private HttpResponse allowGcpArchiveAccess(String tenantName, HttpRequest request) {
+        if (controller.tenants().require(TenantName.from(tenantName)).type() != Tenant.Type.cloud)
+            throw new IllegalArgumentException("Tenant '" + tenantName + "' is not a cloud tenant");
+
+        var data = toSlime(request.getData()).get();
+        var member = mandatory("member", data).asString();
+
+        if (member.isBlank()) {
+            return ErrorResponse.badRequest("GCP archive access role can't be whitespace only");
+        }
+
+        controller.tenants().lockOrThrow(TenantName.from(tenantName), LockedTenant.Cloud.class, lockedTenant -> {
+            var access = lockedTenant.get().archiveAccess();
+            lockedTenant = lockedTenant.withArchiveAccess(access.withGCPMember(member));
+            controller.tenants().store(lockedTenant);
+        });
+
+        return new MessageResponse("GCP archive access member set to '" + member + "' for tenant " + tenantName +  ".");
+    }
+
+    private HttpResponse removeGcpArchiveAccess(String tenantName) {
+        if (controller.tenants().require(TenantName.from(tenantName)).type() != Tenant.Type.cloud)
+            throw new IllegalArgumentException("Tenant '" + tenantName + "' is not a cloud tenant");
+
+        controller.tenants().lockOrThrow(TenantName.from(tenantName), LockedTenant.Cloud.class, lockedTenant -> {
+            var access = lockedTenant.get().archiveAccess();
+            lockedTenant = lockedTenant.withArchiveAccess(access.removeGCPMember());
+            controller.tenants().store(lockedTenant);
+        });
+
+        return new MessageResponse("GCP archive access member removed for tenant " + tenantName + ".");
     }
 
     private HttpResponse patchApplication(String tenantName, String applicationName, HttpRequest request) {
@@ -2354,7 +2393,9 @@ public class ApplicationApiHandler extends AuditLoggingRequestHandler {
                     log.warning(String.format("Failed to get quota for tenant %s: %s", tenant.name(), Exceptions.toMessageString(e)));
                 }
 
+                // TODO(enygaard, 2022-05-25) Remove when console is using new archive access structure
                 cloudTenant.archiveAccess().awsRole().ifPresent(role -> object.setString("archiveAccessRole", role));
+                toSlime(cloudTenant.archiveAccess(), object.setObject("archiveAccess"));
 
                 break;
             }
@@ -2383,6 +2424,11 @@ public class ApplicationApiHandler extends AuditLoggingRequestHandler {
             }
         }
         tenantMetaDataToSlime(tenant, applications, object.setObject("metaData"));
+    }
+
+    private void toSlime(ArchiveAccess archiveAccess, Cursor object) {
+        archiveAccess.awsRole().ifPresent(role -> object.setString("awsRole", role));
+        archiveAccess.gcpMember().ifPresent(member -> object.setString("gcpMember", member));
     }
 
     private void toSlime(Quota quota, QuotaUsage usage, Cursor object) {

--- a/controller-server/src/test/java/com/yahoo/vespa/hosted/controller/maintenance/ArchiveAccessMaintainerTest.java
+++ b/controller-server/src/test/java/com/yahoo/vespa/hosted/controller/maintenance/ArchiveAccessMaintainerTest.java
@@ -45,8 +45,8 @@ public class ArchiveAccessMaintainerTest {
         assertEquals(0, archiveService.authorizeAccessByTenantName.size());
         MockMetric metric = new MockMetric();
         new ArchiveAccessMaintainer(tester.controller(), metric, Duration.ofMinutes(10)).maintain();
-        assertEquals(new ArchiveAccess(Optional.of(tenant1role), Optional.empty()), archiveService.authorizeAccessByTenantName.get(tenant1));
-        assertEquals(new ArchiveAccess(Optional.of(tenant2role), Optional.empty()), archiveService.authorizeAccessByTenantName.get(tenant2));
+        assertEquals(new ArchiveAccess().withAWSRole(tenant1role), archiveService.authorizeAccessByTenantName.get(tenant1));
+        assertEquals(new ArchiveAccess().withAWSRole(tenant2role), archiveService.authorizeAccessByTenantName.get(tenant2));
 
         var expected = Map.of("archive.bucketCount",
                               tester.controller().zoneRegistry().zonesIncludingSystem().all().ids().stream()
@@ -60,7 +60,7 @@ public class ArchiveAccessMaintainerTest {
     private TenantName createTenantWithAccessRole(ControllerTester tester, String tenantName, String role) {
         var tenant = tester.createTenant(tenantName, Tenant.Type.cloud);
         tester.controller().tenants().lockOrThrow(tenant, LockedTenant.Cloud.class, lockedTenant -> {
-            lockedTenant = lockedTenant.withArchiveAccess(new ArchiveAccess(Optional.of(role), Optional.empty()));
+            lockedTenant = lockedTenant.withArchiveAccess(new ArchiveAccess().withAWSRole(role));
             tester.controller().tenants().store(lockedTenant);
         });
         return tenant;

--- a/controller-server/src/test/java/com/yahoo/vespa/hosted/controller/persistence/TenantSerializerTest.java
+++ b/controller-server/src/test/java/com/yahoo/vespa/hosted/controller/persistence/TenantSerializerTest.java
@@ -125,7 +125,7 @@ public class TenantSerializerTest {
                         new TenantSecretStore("ss1", "123", "role1"),
                         new TenantSecretStore("ss2", "124", "role2")
                 ),
-                new ArchiveAccess(Optional.of("arn:aws:iam::123456789012:role/my-role"), Optional.empty())
+                new ArchiveAccess().withAWSRole("arn:aws:iam::123456789012:role/my-role")
         );
         CloudTenant serialized = (CloudTenant) serializer.tenantFrom(serializer.toSlime(tenant));
         assertEquals(tenant.info(), serialized.info());
@@ -174,7 +174,7 @@ public class TenantSerializerTest {
                         otherPublicKey, new SimplePrincipal("jane")),
                 TenantInfo.empty(),
                 List.of(),
-                new ArchiveAccess(Optional.of("arn:aws:iam::123456789012:role/my-role"), Optional.of("user:foo@example.com"))
+                new ArchiveAccess().withAWSRole("arn:aws:iam::123456789012:role/my-role").withGCPMember("user:foo@example.com")
         );
         CloudTenant serialized = (CloudTenant) serializer.tenantFrom(serializer.toSlime(tenant));
         assertEquals(serialized.archiveAccess().awsRole().get(), "arn:aws:iam::123456789012:role/my-role");

--- a/controller-server/src/test/java/com/yahoo/vespa/hosted/controller/restapi/application/ApplicationApiCloudTest.java
+++ b/controller-server/src/test/java/com/yahoo/vespa/hosted/controller/restapi/application/ApplicationApiCloudTest.java
@@ -315,9 +315,40 @@ public class ApplicationApiCloudTest extends ControllerContainerCloudTest {
                 .data("{\"role\":\"dummy\"}").roles(Role.administrator(tenantName)),
                 "{\"error-code\":\"BAD_REQUEST\",\"message\":\"Invalid archive access role 'dummy': Must match expected pattern: 'arn:aws:iam::\\\\d{12}:.+'\"}", 400);
 
+        tester.assertResponse(request("/application/v4/tenant/scoober/archive-access/aws", PUT)
+                        .data("{\"role\":\"arn:aws:iam::123456789012:role/my-role\"}").roles(Role.administrator(tenantName)),
+                "{\"message\":\"AWS archive access role set to 'arn:aws:iam::123456789012:role/my-role' for tenant scoober.\"}", 200);
+        tester.assertResponse(request("/application/v4/tenant/scoober", GET).roles(Role.reader(tenantName)),
+                (response) -> assertTrue(response.getBodyAsString().contains("\"archiveAccessRole\":\"arn:aws:iam::123456789012:role/my-role\"")),
+                200);
+        tester.assertResponse(request("/application/v4/tenant/scoober/archive-access/aws", DELETE).roles(Role.administrator(tenantName)),
+                "{\"message\":\"AWS archive access role removed for tenant scoober.\"}", 200);
+        tester.assertResponse(request("/application/v4/tenant/scoober", GET).roles(Role.reader(tenantName)),
+                (response) -> assertFalse(response.getBodyAsString().contains("\"archiveAccessRole\":\"arn:aws:iam::123456789012:role/my-role\"")),
+                200);
+
+        tester.assertResponse(request("/application/v4/tenant/scoober/archive-access/gcp", PUT)
+                        .data("{\"member\":\"user:test@example.com\"}").roles(Role.administrator(tenantName)),
+                "{\"message\":\"GCP archive access member set to 'user:test@example.com' for tenant scoober.\"}", 200);
+        tester.assertResponse(request("/application/v4/tenant/scoober", GET).roles(Role.reader(tenantName)),
+                (response) -> assertTrue(response.getBodyAsString().contains("\"gcpMember\":\"user:test@example.com\"")),
+                200);
+        tester.assertResponse(request("/application/v4/tenant/scoober/archive-access/gcp", DELETE).roles(Role.administrator(tenantName)),
+                "{\"message\":\"GCP archive access member removed for tenant scoober.\"}", 200);
+        tester.assertResponse(request("/application/v4/tenant/scoober", GET).roles(Role.reader(tenantName)),
+                (response) -> assertFalse(response.getBodyAsString().contains("\"gcpMember\":\"user:test@example.com\"")),
+                200);
+
         tester.assertResponse(request("/application/v4/tenant/scoober/archive-access", PUT)
                         .data("{\"role\":\"arn:aws:iam::123456789012:role/my-role\"}").roles(Role.administrator(tenantName)),
-                "{\"message\":\"Archive access role set to 'arn:aws:iam::123456789012:role/my-role' for tenant scoober.\"}", 200);
+                "{\"message\":\"AWS archive access role set to 'arn:aws:iam::123456789012:role/my-role' for tenant scoober.\"}", 200);
+        tester.assertResponse(request("/application/v4/tenant/scoober", GET).roles(Role.reader(tenantName)),
+                (response) -> assertTrue(response.getBodyAsString().contains("\"archiveAccessRole\":\"arn:aws:iam::123456789012:role/my-role\"")),
+                200);
+
+        tester.assertResponse(request("/application/v4/tenant/scoober/archive-access", PUT)
+                        .data("{\"role\":\"arn:aws:iam::123456789012:role/my-role\"}").roles(Role.administrator(tenantName)),
+                "{\"message\":\"AWS archive access role set to 'arn:aws:iam::123456789012:role/my-role' for tenant scoober.\"}", 200);
         tester.assertResponse(request("/application/v4/tenant/scoober", GET).roles(Role.reader(tenantName)),
                 (response) -> assertTrue(response.getBodyAsString().contains("\"archiveAccessRole\":\"arn:aws:iam::123456789012:role/my-role\"")),
                 200);
@@ -327,7 +358,7 @@ public class ApplicationApiCloudTest extends ControllerContainerCloudTest {
                 new File("deployment-cloud.json"));
 
         tester.assertResponse(request("/application/v4/tenant/scoober/archive-access", DELETE).roles(Role.administrator(tenantName)),
-                "{\"message\":\"Archive access role removed for tenant scoober.\"}", 200);
+                "{\"message\":\"AWS archive access role removed for tenant scoober.\"}", 200);
         tester.assertResponse(request("/application/v4/tenant/scoober", GET).roles(Role.reader(tenantName)),
                 (response) -> assertFalse(response.getBodyAsString().contains("archiveAccessRole")),
                 200);

--- a/controller-server/src/test/java/com/yahoo/vespa/hosted/controller/restapi/user/responses/tenant-with-keys.json
+++ b/controller-server/src/test/java/com/yahoo/vespa/hosted/controller/restapi/user/responses/tenant-with-keys.json
@@ -24,6 +24,7 @@
     "budgetUsed": 0.0,
     "clusterSize": 5
   },
+  "archiveAccess": { },
   "applications": [
     {
       "tenant": "my-tenant",

--- a/controller-server/src/test/java/com/yahoo/vespa/hosted/controller/restapi/user/responses/tenant-with-secrets.json
+++ b/controller-server/src/test/java/com/yahoo/vespa/hosted/controller/restapi/user/responses/tenant-with-secrets.json
@@ -32,6 +32,7 @@
     "budgetUsed": 0.0,
     "clusterSize": 5
   },
+  "archiveAccess": { },
   "applications": [
     {
       "tenant": "my-tenant",

--- a/controller-server/src/test/java/com/yahoo/vespa/hosted/controller/restapi/user/responses/tenant-without-applications.json
+++ b/controller-server/src/test/java/com/yahoo/vespa/hosted/controller/restapi/user/responses/tenant-without-applications.json
@@ -15,6 +15,7 @@
     "budgetUsed": 0.0,
     "clusterSize": 5
   },
+  "archiveAccess": { },
   "applications": [ ],
   "metaData": {
     "createdAtMillis": 1600000000000

